File: backend/src/CHIS.Core/Entities/Immunization.cs
namespace CHIS.Core.Entities;

public class Antigen : BaseEntity
{
    public string Code { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
    public string? Description { get; set; }
    public bool IsActive { get; set; } = true;
}

public class Vaccine : BaseEntity
{
    public string Code { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
    public string? Manufacturer { get; set; }
    public string? AntigenList { get; set; }
    public string? StorageCondition { get; set; }
    public int? DosesPerVial { get; set; }
    public bool IsActive { get; set; } = true;
}

public class ImmunizationSubject : BaseEntity
{
    public string SubjectCode { get; set; } = string.Empty;
    public string FullName { get; set; } = string.Empty;
    public DateTime DateOfBirth { get; set; }
    public int Gender { get; set; }
    public string? MotherName { get; set; }
    public string? FatherName { get; set; }
    public string? Address { get; set; }
    public string? Village { get; set; }
    public string? Phone { get; set; }
    public Guid? FacilityId { get; set; }
    public ICollection<VaccinationRecord> Vaccinations { get; set; } = new List<VaccinationRecord>();
    public ICollection<NutritionMeasurement> NutritionMeasurements { get; set; } = new List<NutritionMeasurement>();
}

public class VaccinationRecord : BaseEntity
{
    public Guid SubjectId { get; set; }
    public Guid VaccineId { get; set; }
    public DateTime VaccinationDate { get; set; }
    public int DoseNumber { get; set; }
    public string? BatchNumber { get; set; }
    public string? InjectionSite { get; set; }
    public string? Route { get; set; }
    public string? Reaction { get; set; }
    public string? ReactionDetail { get; set; }
    public Guid? VaccinatorId { get; set; }
    public Guid? FacilityId { get; set; }
    public ImmunizationSubject Subject { get; set; } = null!;
    public Vaccine Vaccine { get; set; } = null!;
}

public class ImmunizationPlan : BaseEntity
{
    public string PlanCode { get; set; } = string.Empty;
    public int Year { get; set; }
    public int Month { get; set; }
    public Guid FacilityId { get; set; }
    public string? PlanData { get; set; }
    public int Status { get; set; }
}

public class VaccineStock : BaseEntity
{
    public string StockCode { get; set; } = string.Empty;
    public Guid VaccineId { get; set; }
    public string? StockType { get; set; }
    public DateTime StockDate { get; set; }
    public decimal Quantity { get; set; }
    public string? BatchNumber { get; set; }
    public DateTime? ExpiryDate { get; set; }
    public string? Notes { get; set; }
    public int Status { get; set; }
    public Guid? FacilityId { get; set; }
    public Vaccine Vaccine { get; set; } = null!;
}

public class NutritionMeasurement : BaseEntity
{
    public Guid SubjectId { get; set; }
    public DateTime MeasurementDate { get; set; }
    public decimal? Weight { get; set; }
    public decimal? Height { get; set; }
    public decimal? HeadCircumference { get; set; }
    public string? NutritionStatus { get; set; }
    public string? Notes { get; set; }
    public ImmunizationSubject Subject { get; set; } = null!;
}

File: frontend/src/api/immunization.ts
import client from './client';

export interface ImmunizationSubjectDto {
  id: string;
  subjectCode: string;
  fullName: string;
  dateOfBirth: string;
  gender: number;
  motherName?: string;
  address?: string;
  village?: string;
  phone?: string;
  vaccinationCount: number;
}

export interface CreateImmunizationSubjectDto {
  fullName: string;
  dateOfBirth: string;
  gender: number;
  motherName?: string;
  fatherName?: string;
  address?: string;
  village?: string;
  phone?: string;
}

export interface VaccinationRecordDto {
  id: string;
  subjectId: string;
  subjectName?: string;
  vaccineId: string;
  vaccineName?: string;
  vaccinationDate: string;
  doseNumber: number;
  batchNumber?: string;
  injectionSite?: string;
  route?: string;
  reaction?: string;
  reactionDetail?: string;
}

export interface CreateVaccinationRecordDto {
  subjectId: string;
  vaccineId: string;
  doseNumber: number;
  batchNumber?: string;
  injectionSite?: string;
  route?: string;
}

export interface VaccineDto {
  id: string;
  code: string;
  name: string;
  manufacturer?: string;
  antigenList?: string;
  storageCondition?: string;
  dosesPerVial?: number;
  isActive: boolean;
}

export interface VaccineStockDto {
  id: string;
  stockCode: string;
  vaccineId: string;
  vaccineName?: string;
  stockType?: string;
  stockDate: string;
  quantity: number;
  batchNumber?: string;
  expiryDate?: string;
}

export interface NutritionMeasurementDto {
  id: string;
  subjectId: string;
  subjectName?: string;
  measurementDate: string;
  weight?: number;
  height?: number;
  headCircumference?: number;
  nutritionStatus?: string;
}

export interface VaccineStockIssueDto {
  id: string;
  issueCode?: string;
  issueType: string;
  vaccineId: string;
  vaccineName?: string;
  quantity: number;
  issueDate: string;
  reason?: string;
  batchNumber?: string;
  notes?: string;
  status: number;
}

export interface CreateVaccineStockIssueDto {
  issueType: string;
  vaccineId: string;
  quantity: number;
  reason?: string;
  batchNumber?: string;
  notes?: string;
}

export interface ImmunReportDto {
  id: string;
  reportCode: string;
  year: number;
  month?: number;
  quarter?: number;
  facilityId?: string;
  facilityName?: string;
  data?: string;
  status: number;
  sentDate?: string;
}

export interface ChildAgeStatsDto {
  under1Year: number;
  from1To2Years: number;
  from2To5Years: number;
  above5Years: number;
  totalSubjects: number;
  fullyVaccinated: number;
  partiallyVaccinated: number;
}

export const immunizationApi = {
  // Subjects
  getSubjects: (params: { keyword?: string; village?: string; pageIndex?: number; pageSize?: number }) =>
    client.get<{ items: ImmunizationSubjectDto[]; totalCount: number }>('/immunization/subjects', { params }),
  getSubjectById: (id: string) => client.get<ImmunizationSubjectDto>(`/immunization/subjects/${id}`),
  createSubject: (data: CreateImmunizationSubjectDto) =>
    client.post('/immunization/subjects', data),

  // Vaccinations
  getVaccinations: (subjectId: string) =>
    client.get<VaccinationRecordDto[]>(`/immunization/subjects/${subjectId}/vaccinations`),
  recordVaccination: (data: CreateVaccinationRecordDto) =>
    client.post('/immunization/vaccinations', data),

  // Vaccines
  getVaccines: () => client.get<VaccineDto[]>('/immunization/vaccines'),
  createVaccine: (data: VaccineDto) =>
    client.post('/immunization/vaccines', data),

  // Vaccine stock
  getVaccineStock: (vaccineId?: string) =>
    client.get<VaccineStockDto[]>('/immunization/vaccine-stock', { params: { vaccineId } }),

  // Nutrition
  getNutritionMeasurements: (subjectId: string) =>
    client.get<NutritionMeasurementDto[]>(`/immunization/subjects/${subjectId}/nutrition`),
  recordMeasurement: (subjectId: string, data: NutritionMeasurementDto) =>
    client.post(`/immunization/subjects/${subjectId}/nutrition`, data),

  // Vaccine stock issues
  createVaccineStockIssue: (data: CreateVaccineStockIssueDto) =>
    client.post<VaccineStockIssueDto>('/immunization/vaccine-stock-issues', data),
  getVaccineStockIssues: (params?: { issueType?: string; pageIndex?: number; pageSize?: number }) =>
    client.get<{ items: VaccineStockIssueDto[]; totalCount: number }>('/immunization/vaccine-stock-issues', { params }),

  // Reports
  getReport: (reportCode: string, params: { year: number; month?: number; quarter?: number; facilityId?: string }) =>
    client.get<ImmunReportDto>(`/immunization/reports/${reportCode}`, { params }),
  sendReport: (reportCode: string, reportId: string) =>
    client.post(`/immunization/reports/${reportCode}/${reportId}/send`),

  // Print
  printBarcode: (subjectId: string) =>
    client.get(`/immunization/subjects/${subjectId}/barcode`, { responseType: 'blob' }),
  printAppointmentSlip: (planId: string) =>
    client.get(`/immunization/plans/${planId}/appointment-slip`, { responseType: 'blob' }),

  // Child stats
  getChildStats: (facilityId?: string) =>
    client.get<ChildAgeStatsDto>('/immunization/child-stats', { params: { facilityId } }),
};
